File: console.py
import datetime
from pathlib import Path

class Console:
    console = []
    consoleExists = False

    @staticmethod
    def get_timestamp():
        return datetime.datetime.now().strftime("%m/%d/%Y %H:%M")

    @classmethod
    def new(cls, object_type):
        if object_type == "console":
            cls.console = []
            cls.consoleExists = True
            print(f"[{cls.get_timestamp()}] Console object created/reset.")
        elif object_type == "log":
            print(f"[{cls.get_timestamp()}] Log object type not yet implemented.")
        else:
            print(f"[{cls.get_timestamp()}] ERROR! Invalid object type.")

    @classmethod
    def log(cls, msg):
        if cls.consoleExists:
            timestamped_msg = f"[{cls.get_timestamp()}] {msg}"
            cls.console.append(timestamped_msg)
            print(f"Logged: '{timestamped_msg}'")
        else:
            print(f"[{cls.get_timestamp()}] ERROR! No existing console object.")

    @classmethod
    def terminate(cls, index):
        if cls.consoleExists:
            try:
                removed_item = cls.console.pop(index)
                cls.log(f"Terminated line {index}: '{removed_item.strip()}'")
            except IndexError:
                cls.log(f"ERROR! Index {index} is out of range.")
        else:
            print(f"[{cls.get_timestamp()}] ERROR! No existing console object.")

    @classmethod
    def clear(cls):
        if cls.consoleExists:
            cls.console = []
            cls.log("Console cleared.")
        else:
            print(f"[{cls.get_timestamp()}] ERROR! No existing console object.")

    @classmethod
    def println(cls, index):
        if cls.consoleExists:
            try:
                print(cls.console[index])
            except IndexError:
                print(f"[{cls.get_timestamp()}] ERROR! Index {index} is out of range.")
        else:
            print(f"[{cls.get_timestamp()}] ERROR! No existing console object.")

    @classmethod
    def prntall(cls):
        if cls.consoleExists:
            if not cls.console:
                cls.log("Console is empty.")
                return
            print("--- Console Start ---")
            for item in cls.console:
                print(item)
            print("--- Console End ---")
        else:
            print(f"[{cls.get_timestamp()}] ERROR! No existing console object.")

    @classmethod
    def export_log(cls, filename="output_log", directory="."):
        if cls.consoleExists and cls.console:
            try:
                file_path = Path(directory) / f"{filename}.bmlog"
                file_path.parent.mkdir(parents=True, exist_ok=True)
                with open(file_path, "w") as f:
                    for line in cls.console:
                        f.write(f"{line}\n")
                cls.log(f"Successfully exported console to '{file_path}'.")
            except IOError as e:
                cls.log(f"ERROR! Could not write to file {file_path}: {e}")
        elif cls.consoleExists and not cls.console:
            cls.log("Console is empty; nothing to export.")
        else:
            cls.log("ERROR! No existing console object to export.")

File: console.js
const fs = require("fs");
const path = require("path");

const Console = {
    console: [],
    consoleExists: false,

    getTimestamp() {
        const now = new Date();
        const m = String(now.getMonth() + 1).padStart(2, "0");
        const d = String(now.getDate()).padStart(2, "0");
        const y = now.getFullYear();
        const h = String(now.getHours()).padStart(2, "0");
        const min = String(now.getMinutes()).padStart(2, "0");
        return `${m}/${d}/${y} ${h}:${min}`;
    },

    new(objectType) {
        if (objectType === "console") {
            this.console = [];
            this.consoleExists = true;
            console.log(`[${this.getTimestamp()}] Console object created/reset.`);
        } else if (objectType === "log") {
            console.log(`[${this.getTimestamp()}] Log object type not yet implemented.`);
        } else {
            console.log(`[${this.getTimestamp()}] ERROR! Invalid object type.`);
        }
    },

    log(msg) {
        if (this.consoleExists) {
            const timestamped = `[${this.getTimestamp()}] ${msg}`;
            this.console.push(timestamped);
            console.log(`Logged: '${timestamped}'`);
        } else {
            console.log(`[${this.getTimestamp()}] ERROR! No existing console object.`);
        }
    },

    terminate(index) {
        if (this.consoleExists) {
            if (index >= 0 && index < this.console.length) {
                const removed = this.console.splice(index, 1)[0];
                this.log(`Terminated line ${index}: '${removed}'`);
            } else {
                this.log(`ERROR! Index ${index} is out of range.`);
            }
        } else {
            console.log(`[${this.getTimestamp()}] ERROR! No existing console object.`);
        }
    },

    clear() {
        if (this.consoleExists) {
            this.console = [];
            this.log("Console cleared.");
        } else {
            console.log(`[${this.getTimestamp()}] ERROR! No existing console object.`);
        }
    },

    println(index) {
        if (this.consoleExists) {
            if (index >= 0 && index < this.console.length) {
                console.log(this.console[index]);
            } else {
                console.log(`[${this.getTimestamp()}] ERROR! Index ${index} is out of range.`);
            }
        } else {
            console.log(`[${this.getTimestamp()}] ERROR! No existing console object.`);
        }
    },

    prntall() {
        if (this.consoleExists) {
            if (this.console.length === 0) {
                this.log("Console is empty.");
                return;
            }
            console.log("--- Console Start ---");
            this.console.forEach(line => console.log(line));
            console.log("--- Console End ---");
        } else {
            console.log(`[${this.getTimestamp()}] ERROR! No existing console object.`);
        }
    },

    exportLog(filename = "output_log", directory = ".") {
        if (this.consoleExists && this.console.length > 0) {
            const dirPath = path.resolve(directory);
            if (!fs.existsSync(dirPath)) fs.mkdirSync(dirPath, { recursive: true });
            const filePath = path.join(dirPath, `${filename}.bmlog`);
            try {
                fs.writeFileSync(filePath, this.console.join("\n"), "utf8");
                this.log(`Successfully exported console to '${filePath}'.`);
            } catch (e) {
                this.log(`ERROR! Could not write to file ${filePath}: ${e}`);
            }
        } else if (this.consoleExists) {
            this.log("Console is empty; nothing to export.");
        } else {
            this.log("ERROR! No existing console object to export.");
        }
    }
};

module.exports = Console;
